pkg/process/heartbeat/monitor.go

```diff
@@ -0,0 +1,172 @@
+package heartbeat
+
+import (
+	"fmt"
+	"sort"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/process/net"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-go/statsd"
+	"k8s.io/apimachinery/pkg/util/sets"
+)
+
+// MetricName is used to determine the metric name for a given module
+func MetricName(moduleName string) string {
+	return fmt.Sprintf("datadog.system_probe.agent.%s", moduleName)
+}
+
+// ModuleMonitor is responsible for emitting heartbeat metrics for each
+// system-probe module. It does so by hitting the stats endpoint from
+// system-probe and emitting one metric per enabled module using Datadog Public
+// API.
+type ModuleMonitor struct {
+	enabledModulesFn func() ([]string, error)
+	metricNameFn     func(string) string
+	flusher          flusher
+	exit             chan struct{}
+}
+
+// Options encapsulates all configuration params used by ModuleMonitor
+type Options struct {
+	// KeysPerDomain (required) contains API key entries per Datadog domain
+	KeysPerDomain map[string][]string
+
+	// HostName (required)
+	HostName string
+
+	// SysprobeSocketPath (optional) sets the location of the Unix socket used
+	// to reach system-probe
+	SysprobeSocketPath string
+
+	// StatsdClient (optional) points to a statsd client to be used as a
+	// fallback mechanism
+	StatsdClient statsd.ClientInterface
+
+	// TagVersion (optional) contains the agent version to be sent along with the
+	// hearbeat metric
+	TagVersion string
+
+	// TagRevision (optional) contains the agent revision to be sent along with the
+	// hearbeat metric
+	TagRevision string
+
+	// MetricNameFn (optional) allows metric names to be specified
+	MetricNameFn func(string) string
+}
+
+// NewModuleMonitor returns a new ModuleMonitor
+func NewModuleMonitor(opts Options) (*ModuleMonitor, error) {
+	if opts.SysprobeSocketPath != "" {
+		net.SetSystemProbePath(opts.SysprobeSocketPath)
+	}
```

Comment:
Is this necessary? I'm not a fan of setting this globally in multiple places.

Reply:
Me neither, but this is necessary if the socket path is not the default one. In the context of `process-agent` I believe I can remove the line that sets `opts.SysprobeSocketPath` though.

Reply:
We could also pass in an instance of  RemoteSysProbeUtil  and leave the initialization to whatever initializes this.

Reply:
@leeavital I tried to make that change, but unfortunately `GetRemoteSystemProbeUtil()` returns `nil` instead of a `RemoteSysProbeUtil` instance until system-probe is initialized. 🤷‍♂️ 
That means that clients of this library would have to wait system-probe to initialize before calling `heartbeat.NewMonitor`.